core/include/thread_flags.h

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (C) 2015 Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     core_thread
+ * @brief       Thread flags
+ * @{
+ *
+ * @file
+ * @brief       Thread flags API
+ *
+ * This API can be used to notify threads of conditions in a race-free
+ * and allocation-less way.
+ *
+ * Each thread has a number boolean flags (usually 16) in it's tcb, stored in
+ * it's flags field. Those flags can be set or unset, using thread_flags_set(),
+ * from ISR's, other threads or even by the thread itself.
+ *
+ * A thread can wait for any combination of it's flags to become set, using
+ * thread_flags_wait_any() or thread_flags_wait_all().
+ * Those functions clear flags that caused them to return.
+ * It is not possible to wait for flags to become unset.
+ *
+ * Thread flags, apart from being very efficient, have certain properties
+ * that make them the preferred choice over messages or mutexes in some
+ * circumstances:
+ *
+ * - setting thread flags cannot fail
+ *   If messages are used to notify a thread of a condition from within an ISR,
+ *   and the receiving thread is not waiting, has no queue or the queue is
+ *   full, the ISR cannot deliver the message. A thread flag can always be set.
```

Comment:
Should we mention the downside or rather the use cases where messages must be used (e.g. if it is not only of interest **that** an event occurred, but **how many** of them occurred?